training_controller/prepare_training_logs.py

```diff
@@ -175,78 +147,126 @@ def fetch_training_logs(self, es_instance, num_logs_to_fetch, timestamps_list):
         oldest_log_timestamp = int(oldest_log["aggregations"]["min_ts"]["value"])
+        newest_log_timestamp = int(newest_log["aggregations"]["max_ts"]["value"])
         try:
             all_normal_intervals = scan(es_instance, index="opni-normal-intervals", query={"query": {"match_all": {}}})
         except Exception as e:
             logging.error("Error trying to retrieve all normal intervals from opni-normal-intervals index")
             return timestamps_list
         for normal_interval in all_normal_intervals:
             start_ts, end_ts = normal_interval["_source"]["start_ts"], normal_interval["_source"]["end_ts"]
+            full_file_prefix = "{}_{}".format(start_ts, end_ts)
+            interval_json_files = [
+                file
+                for file in os.listdir(self.TRAINING_DIR)
+                if full_file_prefix in file
+            ]
             if end_ts < oldest_log_timestamp:
                 try:
                     es_instance.delete(index="opni-normal-intervals", doc_type=normal_interval["_type"], id=normal_interval["_id"])
                     logging.info("Deleting old normal time interval from Elasticsearch")
+                    for interval_file in interval_json_files:
+                        os.remove(os.path.join(self.TRAINING_DIR,interval_file))
+                        logging.info("Removing old JSON training files where the time interval within opni-normal-intervals was deleted.")
                 except Exception as e:
                     logging.error("Error deleting document from opni-normal-intervals index.")
                     continue
             elif start_ts < oldest_log_timestamp:
-                timestamps_list.append({"start_ts": oldest_log_timestamp, "end_ts": end_ts})
+                if end_ts > newest_log_timestamp:
+                    timestamps_list.append({"start_ts": oldest_log_timestamp, "end_ts": end_ts,"filename": "{}_{}.json".format(oldest_log_timestamp, newest_log_timestamp)})
+                else:
+                    timestamps_list.append({"start_ts": oldest_log_timestamp, "end_ts": end_ts, "filename": "{}_{}.json".format(oldest_log_timestamp, end_ts)})
                 try:
                     es_instance.update(index="opni-normal-intervals", doc_type=normal_interval["_type"], id=normal_interval["_id"], body={"doc": {"start_ts": oldest_log_timestamp}})
                     logging.info("Updating time interval within Elasticsearch.")
+                    for interval_file in interval_json_files:
+                        logging.info("Removing old JSON training files where the time interval within opni-normal-intervals was updated.")
+                        os.remove(os.path.join(self.TRAINING_DIR, interval_file))
                 except Exception as e:
                     logging.error("Error updating document within opni-normal-intervals index.")
                     continue
+            elif end_ts > newest_log_timestamp:
```

Comment:
why would this case ever occurs?  isn't the `end_ts` generated by `drain_service` and it should never be newer than `newest_log_timestamp`?

Reply:
I was thinking more in an event where there is a delay in sending those logs over to Elasticsearch and thus when that query is made, not all of the logs are placed in Elasticsearch yet. This would be a worst case situation.